### ISSUE TITLE
Unboxability check breaks with module constraints

### ISSUE THREAD

**Issue by stedolan** [ocaml/ocaml#9607]
This program was accepted with 4.10:
```
module type T  = sig type 'k t end

module M : T with type 'k t = string = struct
  type 'k t = string
end

type t = T : 'k M.t -> t [@@unboxed]
```
but with 4.11 or trunk, gives this error:
```
File "/home/sdolan/q/hax/unbox.ml", line 7, characters 0-36:
7 | type t = T : 'k M.t -> t [@@unboxed]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This type cannot be unboxed because
       it might contain both float and non-float values,
       depending on the instantiation of the existential variable 'k.
       You should annotate it with [@@ocaml.boxed].
```
Removing the no-op `with type 'k t = string` constraint makes it pass again.

Curiously, manually expanding `module M : T with type 'k t = string` into `module M : sig type 'k t = string end` makes it pass as well, so the issue is probably something to do with with-constraints?

(cc @rlepigre, @gasche)

**Comment by gasche** [comment]
Thanks for the report! It looks like the checker believes that `'a M.t` in your example is an abstract type. We probably missed a case, in the representation of types, of a type that is abstract but has a definition (manifest?). If this is the issue, it should be very simple to fix.

**Comment by gasche** [comment]
(No, we are correctly handling the abstract-with-manifest case I think, so this is not the issue.)

**Comment by gasche** [comment]
I proposed a fix in #9609.

**Comment by stedolan** [comment]
Thanks for the quick fix!
